cli_tools/google-osconfig-agent/ospatch/patch_run.go

```diff
@@ -277,70 +278,58 @@ func (r *patchRun) reportSucceeded(ctx context.Context) {
  * - An error occurred and we do another attempt starting where we last failed.
  * - The process was unexpectedly restarted and we are continuing from where we left off.
  */
-func (r *patchRun) runPatch(ctx context.Context) {
-	savedPatchState, err := loadState(state)
-	if err != nil {
-		logger.Errorf("loadState error: %v", err)
-		return
-	}
-
-	if savedPatchState != nil && savedPatchState.Job.PatchJobName == r.Job.PatchJobName {
-		// continue from previous patch step
-		r.PatchStep = savedPatchState.PatchStep
-	} else {
-		// We have no saved state for this patch.
-		r.PatchStep = started
+func (r *patchRun) runPatch() {
+	logger.Debugf("Running patch job %s", r.Job.PatchJobName)
+	if err := r.createClient(); err != nil {
+		logger.Errorf("Error creating osconfig client: %v", err)
 	}
+	defer r.close()
 
-	if patchStepIndex[r.PatchStep] <= patchStepIndex[started] {
-		logger.Debugf("Starting patchJob %s", r.Job)
-		r.StartedAt = time.Now()
-
-		if r.reportState(ctx, osconfigpb.Instance_STARTED) {
+	for {
+		switch r.PatchStep {
+		default:
+			r.finishAndReportError(fmt.Sprintf("Unknown step: %q", r.PatchStep))
 			return
-		}
-	}
+		case acked:
+			logger.Debugf("Starting patchJob %s", r.Job)
+			r.StartedAt = time.Now()
+			r.PatchStep = prePatchReboot
 
-	if patchStepIndex[r.PatchStep] <= patchStepIndex[prePatchReboot] {
-		// check if we need to reboot
-		if r.rebootIfNeeded(ctx, false) {
-			return
-		}
-	}
-
-	if patchStepIndex[r.PatchStep] <= patchStepIndex[applyPatch] {
-		if r.Job.DryRun {
-			if r.reportState(ctx, osconfigpb.Instance_APPLYING_PATCHES) {
+			if r.reportState(osconfigpb.Instance_STARTED) {
 				return
 			}
-			logger.Infof("Dry run - No updates applied for patch job '%s'", r.Job.PatchJobName)
-		} else {
-			err = runUpdates(ctx, r)
-			if err != nil {
-				r.finishAndReportError(ctx, fmt.Sprintf("Failed to apply patches: %v", err))
+		case prePatchReboot:
+			if r.prePatchReboot() {
 				return
 			}
-		}
-	}
-
-	if patchStepIndex[r.PatchStep] <= patchStepIndex[postPatchReboot] {
-		// check if we need to reboot
-		if r.rebootIfNeeded(ctx, true) {
+			r.PatchStep = patching
+		case patching:
+			if r.Job.DryRun {
+				if r.reportState(osconfigpb.Instance_APPLYING_PATCHES) {
+					return
+				}
+				logger.Infof("Dry run - No updates applied for patch job '%s'", r.Job.PatchJobName)
+				if r.postPatchReboot() {
+					return
+				}
+				r.PatchStep = completed
+			} else {
+				if err := runUpdates(r); err != nil {
+					r.finishAndReportError(fmt.Sprintf("Failed to apply patches: %v", err))
+					return
+				}
+				if r.postPatchReboot() {
+					return
+				}
+				// We have not rebooted so no need to rerun updates.
+				r.PatchStep = completed
```

Comment:
You can DRY this a little by moving `if r.postPatchReboot() ... r.PatchStep = completed` out of the if/else.